platforms/extensibility/plugin-use/src/main/java/org/gradle/plugin/use/resolve/internal/PluginResolveContext.java

```diff
@@ -16,13 +16,20 @@
 
 package org.gradle.plugin.use.resolve.internal;
 
-import org.gradle.api.internal.plugins.PluginImplementation;
-import org.gradle.plugin.use.PluginId;
+import org.gradle.api.artifacts.Dependency;
 
+/**
+ * Collects dependencies required to apply plugins.
+ */
 public interface PluginResolveContext {
-    void addLegacy(PluginId pluginId, Object dependencyNotation);
 
-    void add(PluginImplementation<?> plugin);
+    /**
+     * Provide a dependency that will be resolved as part of the script classpath.
+     */
+    void visitDependency(Dependency dependency);
```

Comment:
Maybe 'track' as a prefix?

Reply:
Re-reading, maybe "it isn't visiting anything" isn't completely true.

I still think it's much clearer to say add/track/record.  This definitely isn't the Visitor pattern, and I think that suffix and the visit prefix should be reserved.  

There's only one implementation here, so why not just call it `Default`?  And since there's only one implementation, instead of `visit`, which could be doing anything, just have the methods say what they do, like in the interface's javadoc - they collect (add/track/record) dependencies into this context.

Reply:
What would you rename `PluginResolution#visitDependencies` to?

Reply:
I still am struggling to see how `PluginResolveContext` is not a visitor. The `PluginDependencyVisitor` in that case is a "Collecting" implementation of that visitor. 

From the wiki page on the visitor pattern, the Java example uses this class as the visitor interface, which looks very similar to this:
```
interface CarElementVisitor {
    void visit(Body body);
    void visit(Car car);
    void visit(Engine engine);
    void visit(Wheel wheel);
}
```

IMO the current name ("Context") is worse than Visitor. What does this contexualize here? It does not provide any relevant information. What else would you name this thing?

Reply:
Oh, I thought you liked `PluginResolveContext` and that was why you left it.  I'm not a fan of the name either but don't have any great ideas.  PluginResolveEnvironment?  Background?  Influences?  `PluginResolutionRecord` maybe?  IDK.

Reply:
My objection to calling this a Visitor is that there's no attempt or need to add new operations without modifying an object structure here - which is the purpose and value of a Visitor.  There's only one `PluginResolveContext` implementation, and it didn't look like there would ever be a need for more.  

We're not traversing some polymorphic collection of objects, either the `ArtifactRepositoriesPluginResolver`, the `InjectedClasspathPluginResolution` or the `LocalPluginResolution` is calling a method to record their dependencies in the given "context" object.  If it wasn't named `XYZVisitor` I don't think anyone would see it as an instance of the pattern, which is why I think it should be renamed to not confuse.